1. Load the data set into Python using, e.g., load_wine or genfromtxt, as appropriate. In the case of the USPS dataset, merge the original training and test sets into one dataset.

for wine dataset

In [1]:
from sklearn.datasets import load_wine
wine = load_wine()

for USPS

In [2]:
import numpy as np
train = np.genfromtxt('zip.train')
print(train.shape)
#print(type(train))
test = np.genfromtxt('zip.test')
print(test.shape)

(7291, 257)
(2007, 257)


In [3]:
X_data = np.concatenate((train, test))
X = X_data[:,1:257]
y = X_data[:,0]
X_data.shape
print(X_data.shape)
print(X.shape)
print(y.shape)

(9298, 257)
(9298, 256)
(9298,)



2. Divide the dataset into a training set and a test set. You may use the function train_test_split. Use your birthday in the format DDMM as random_state (omit leading zeros if any).

for USPS

In [4]:
from sklearn.model_selection import train_test_split
X1_train,X1_test,y1_train,y1_test=train_test_split(X,y,random_state = 1602)

print(X1_train.shape)
print(X1_test.shape)
print(y1_train.shape)
print(y1_test.shape)

(6973, 256)
(2325, 256)
(6973,)
(2325,)


for wine dataset

In [5]:

X_train,X_test,y_train,y_test=train_test_split(wine['data'],wine['target'],random_state = 1602)
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(133, 13)
(45, 13)
(133,)
(45,)


3. Using cross-validation and the training set only, estimate the generalization accuracy of the MLPClassifier with the default values of the parameters. You may use the function cross_val_score.

4. Get rid of the warnings by varying a parameter, or parameters, such as max_iter. Estimate the generalization accuracy of the MLPClassifier with the modified values of the parameters. (If there are no warnings, the modified values of the parameters may be identical to the default values.)

for wine dataset

In [6]:
# from sklearn.neural_network import MLPClassifier
# from sklearn.model_selection import cross_val_score
# mlp = MLPClassifier(random_state=1602).fit(X_train,y_train) 
# scores = cross_val_score(mlp, X_train, y_train)
# # compute mean cross-validation accuracy
# score = np.mean(scores)
# print(score)


The above code gives output: 0.2330484330484331
It is commented out to hide the warnings which will be resolved below!

In [7]:
from matplotlib import pyplot as plt 
from sklearn.neural_network import MLPClassifier
mlp = MLPClassifier(random_state=1602,max_iter = 10000).fit(X_train,y_train) 


In [8]:
from sklearn.model_selection import cross_val_score
# perform cross-validation
scores = cross_val_score(mlp, X_train, y_train)
# compute mean cross-validation accuracy
score = np.mean(scores)
print(score)

0.2330484330484331


for USPS

In [9]:
# mlp1 = MLPClassifier(random_state=1602).fit(X1_train,y1_train)
# # perform cross-validation
# scores1 = cross_val_score(mlp1, X1_train, y1_train)
# # compute mean cross-validation accuracy
# score1 = np.mean(scores1)
# print(score1)

The above code gives output: 0.9611367715195179
It is commented out to hide the warnings which will be resolved below!

In [10]:
mlp1 = MLPClassifier(random_state=1602,max_iter = 10000).fit(X1_train,y1_train) 

In [11]:
# perform cross-validation
scores1 = cross_val_score(mlp1, X1_train, y1_train)
# compute mean cross-validation accuracy
score1 = np.mean(scores1)
print(score1)

0.9611367715195179


5. Find the test error rate of the MLPClassifier with the modified values of parameters, compare it with the estimate obtained in the previous task (task 4) for the same values of parameters, and write your observations in a markdown cell of your Jupyter notebook.

for wine dataset

In [12]:
test_error = mlp.score(X_test,y_test)
score2 = (1 - test_error)
print(score2)

0.6222222222222222


for USPS

In [13]:
test_error1 = mlp1.score(X1_test,y1_test)
score3 = (1 - test_error1)
print(score3)

0.024946236559139745


generalization accuracy of the MLPClassifier:-
for wine dataset: 0.2330484330484331
for USPS dataset: 0.9611367715195179
test error rate of the MPLClassifier:-
for wine dataset: 0.6222222222222222
for USPS dataset: 0.024946236559139745

The generalization accuracy of MLPClassifier obtained in Task 4 is approximately equal to 1 - test error rate obtained in task 5.

6. Create a pipeline for MLPClassifier involving data normalization and MLPClassifier, and use grid search and cross-validation to tune parameter alpha for the pipeline, avoiding data snooping and data leakage. You may use the scikit-learn class GridSearchCV. Experiment with different ways of doing normalization (such as StandardScaler, MinMaxScaler, RobustScaler, and Normalizer). Which ways are appropriate for either dataset? (The answer, which should be written in your Jupyter notebook, may depend on the results that you obtain for the next task.)

7. Fit the GridSearchCV object of task 6 to the training set and use it to predict the test labels. Write the resulting test error rate in your Jupyter notebook.

for wine dataset

In [14]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.pipeline import Pipeline
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import GridSearchCV

In [15]:
# standard syntax
pipe = Pipeline([("scaler", MinMaxScaler()), ("mlp", MLPClassifier(random_state=1602,max_iter = 10000))])
param_grid = {'mlp__alpha': [0.001, 0.01, 0.1, 1, 10, 100]}
grid = GridSearchCV(pipe, param_grid=param_grid,n_jobs=-1, cv=5)

In [16]:
grid.fit(X_train,y_train)
pred = grid.predict(X_test)
error_rate = np.mean(pred != y_test)
error_rate

0.0

In [17]:
from sklearn.preprocessing import StandardScaler
pipe1 = Pipeline([("scaler", StandardScaler()), ("mlp", MLPClassifier(random_state=1602,max_iter = 10000))])
param_grid = {'mlp__alpha': [0.001, 0.01, 0.1, 1, 10, 100]}
grid1 = GridSearchCV(pipe1, param_grid=param_grid,n_jobs=-1, cv=5)

In [18]:
grid1.fit(X_train,y_train)
pred1 = grid1.predict(X_test)
error_rate1 = np.mean(pred1 != y_test)
error_rate1

0.0

In [19]:
from sklearn.preprocessing import RobustScaler
pipe2 = Pipeline([("scaler", RobustScaler()), ("mlp", MLPClassifier(random_state=1602,max_iter = 10000))])
param_grid = {'mlp__alpha': [0.001, 0.01, 0.1, 1, 10, 100]}
grid2 = GridSearchCV(pipe2, param_grid=param_grid,n_jobs=-1, cv=5)

In [20]:
grid2.fit(X_train,y_train)
pred2 = grid2.predict(X_test)
error_rate2 = np.mean(pred2 != y_test)
error_rate2

0.0

In [21]:
from sklearn.preprocessing import Normalizer
pipe3 = Pipeline([("scaler", Normalizer()), ("mlp", MLPClassifier(random_state=1602,max_iter = 10000))])
param_grid = {'mlp__alpha': [0.001, 0.01, 0.1, 1, 10, 100]}
grid3 = GridSearchCV(pipe3, param_grid=param_grid,n_jobs=-1, cv=5)

In [22]:
grid3.fit(X_train,y_train)
pred3 = grid3.predict(X_test)
error_rate3 = np.mean(pred3 != y_test)
error_rate3

0.08888888888888889

for USPS

In [23]:
#MinMAxScaler
grid.fit(X1_train,y1_train)
pred4 = grid.predict(X1_test)
error_rate4 = np.mean(pred4 != y1_test)
error_rate4

0.024946236559139787

In [24]:
#StandardScaler
grid1.fit(X1_train,y1_train)
pred5 = grid1.predict(X1_test)
error_rate5 = np.mean(pred5 != y1_test)
error_rate5

0.024086021505376344

In [25]:
#RobustScaler
grid2.fit(X1_train,y1_train)
pred6 = grid2.predict(X1_test)
error_rate6 = np.mean(pred6 != y1_test)
error_rate6

0.03397849462365592

In [26]:
#Normalizer
grid3.fit(X1_train,y1_train)
pred7 = grid3.predict(X1_test)
error_rate7 = np.mean(pred7 != y1_test)
error_rate7


0.025376344086021504

9. An alternative to implementing a cross-conformal predictor is to experiment with a Support Vector Machine. Perform tasks 4–7 for the class SVC in scikit-learn, as described in Lab Worksheet 8, Section 3. The parameters to fit using grid search are C and gamma.

for wine dataset

In [27]:
from sklearn.svm import SVC
svc = SVC()
svc.fit(X_train, y_train)
print("Accuracy on training set:", svc.score(X_train, y_train))
print("Accuracy on test set:", svc.score(X_test, y_test))

Accuracy on training set: 0.706766917293233
Accuracy on test set: 0.5777777777777777


In [28]:
# perform cross-validation
scores_svc = cross_val_score(svc, X_train, y_train)
# compute mean cross-validation accuracy
score_svc = np.mean(scores_svc)
print(score_svc)

0.7068376068376068


In [29]:
test_error_svc = svc.score(X_test,y_test)
score4 = (1 - test_error_svc)
print(score4)

0.4222222222222223


In [30]:
# standard syntax
pipe_short = make_pipeline(MinMaxScaler(), SVC())
param_grid1 = {'svc__C': [0.01, 0.1, 1, 10, 100],
              'svc__gamma': [0.001, 0.01, 0.1, 1, 10, 100]}
grid_svc = GridSearchCV(pipe_short, param_grid=param_grid1, cv=5)


In [31]:
grid_svc.fit(X_train,y_train)
pred_svc = grid_svc.predict(X_test)
error_rate_svc = np.mean(pred_svc != y_test)
error_rate_svc

0.022222222222222223

In [32]:
pipe_short1 = make_pipeline(StandardScaler(), SVC())
param_grid1 = {'svc__C': [0.01, 0.1, 1, 10, 100],
              'svc__gamma': [0.001, 0.01, 0.1, 1, 10, 100]}
grid_svc1 = GridSearchCV(pipe_short1, param_grid=param_grid1, cv=5)

In [33]:
grid_svc1.fit(X_train,y_train)
pred_svc1 = grid_svc1.predict(X_test)
error_rate_svc1 = np.mean(pred_svc1 != y_test)
error_rate_svc1

0.0

In [34]:
pipe_short2 = make_pipeline(RobustScaler(), SVC())
param_grid1 = {'svc__C': [0.01, 0.1, 1, 10, 100],
              'svc__gamma': [0.001, 0.01, 0.1, 1, 10, 100]}
grid_svc2 = GridSearchCV(pipe_short2, param_grid=param_grid1, cv=5)

In [35]:
grid_svc2.fit(X_train,y_train)
pred_svc2 = grid_svc2.predict(X_test)
error_rate_svc2 = np.mean(pred_svc2 != y_test)
error_rate_svc2

0.0

In [36]:
pipe_short3 = make_pipeline(Normalizer(), SVC())
param_grid1 = {'svc__C': [0.01, 0.1, 1, 10, 100],
              'svc__gamma': [0.001, 0.01, 0.1, 1, 10, 100]}
grid_svc3 = GridSearchCV(pipe_short3, param_grid=param_grid1, cv=5)

In [37]:
grid_svc3.fit(X_train,y_train)
pred_svc3 = grid_svc3.predict(X_test)
error_rate_svc3 = np.mean(pred_svc3 != y_test)
error_rate_svc3

0.06666666666666667

for USPS

In [38]:
from sklearn.svm import SVC
svc1 = SVC()
svc1.fit(X1_train, y1_train)
print("Accuracy on training set:", svc1.score(X1_train, y1_train))
print("Accuracy on test set:", svc1.score(X1_test, y1_test))

Accuracy on training set: 0.9892442277355514
Accuracy on test set: 0.9759139784946237


In [39]:
# perform cross-validation
scores_svc_usps = cross_val_score(svc1, X1_train, y1_train)
# compute mean cross-validation accuracy
score_svc_usps = np.mean(scores_svc_usps)
print(score_svc_usps)

0.9685945398353415


In [40]:
test_error_svc_usps = svc1.score(X1_test,y1_test)
score_usps = (1 - test_error_svc_usps)
print(score_usps)

0.02408602150537631


In [41]:
#MinMaxsScaler
grid_svc.fit(X1_train,y1_train)
pred_svc_usps = grid_svc.predict(X1_test)
error_rate_svc_usps = np.mean(pred_svc_usps != y1_test)
error_rate_svc_usps

0.024946236559139787

In [42]:
#StandardScaler
grid_svc1.fit(X1_train,y1_train)
pred_svc_usps1 = grid_svc1.predict(X1_test)
error_rate_svc_usps1 = np.mean(pred_svc_usps1 != y1_test)
error_rate_svc_usps1

0.024946236559139787

In [43]:
#RobustScaler
grid_svc2.fit(X1_train,y1_train)
pred_svc_usps2 = grid_svc2.predict(X1_test)
error_rate_svc_usps2 = np.mean(pred_svc_usps2 != y1_test)
error_rate_svc_usps2

0.07354838709677419

In [44]:
#Normalizer
grid_svc3.fit(X1_train,y1_train)
pred_svc_usps3 = grid_svc3.predict(X1_test)
error_rate_svc_usps3 = np.mean(pred_svc_usps3 != y1_test)
error_rate_svc_usps3

0.01978494623655914